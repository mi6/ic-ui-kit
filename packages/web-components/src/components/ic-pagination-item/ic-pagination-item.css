@import "../../global/normalize.css";

a {
  text-align: center;
}

.item-container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(var(--ic-space-xl) + var(--ic-space-sm));
  height: var(--ic-space-xl);
  margin: var(--ic-space-xxs) var(--ic-space-xxxs);
  cursor: pointer;
  text-decoration: none;
}

button {
  background: none;
  border: none;
  outline: inherit;
  position: relative;
}

.ellipsis {
  pointer-events: none;
  color: var(--ic-color-secondary-text);
}

.item-container:focus {
  box-shadow: var(--ic-border-focus);
  outline: var(--ic-hc-focus-outline);
  border-radius: var(--ic-border-radius);
  transition: var(--ic-easing-transition-fast);
}

.page {
  color: var(--ic-action-default);
}

.simple-current {
  width: calc(var(--ic-space-xxl) + var(--ic-space-md));
  height: var(--ic-space-lg);
  cursor: default;
}

/** DEFAULT STATES **/
.item-container.page:hover:not(.selected):not(:focus) {
  border-radius: var(--ic-border-radius);
  background-color: var(--ic-action-default-bg-hover);
}

.item-container.page:active:not(.selected):not(:focus) {
  background-color: var(--ic-action-default-bg-active);
}

.selected {
  pointer-events: none;
  outline: none;
}

.selected::after {
  content: "";
  height: var(--ic-space-xxs);
  width: 100%;
  position: absolute;
  bottom: 0;
  background-color: var(--ic-action-default);
}

.page-selected {
  color: var(--ic-color-primary-text);
}

/** LIGHT STATES **/
.item-container.page.light:hover:not(.selected):not(:focus) {
  background-color: var(--ic-action-light-bg-hover);
}

.item-container.page.light:active:not(.selected):not(:focus) {
  background-color: var(--ic-action-light-bg-active);
}

ic-typography.light {
  color: var(--ic-color-white-text);
}

.selected.light::after {
  content: "";
  height: var(--ic-space-xxs);
  width: 100%;
  position: absolute;
  bottom: 0;
  background-color: var(--ic-action-light);
}

/** DARK STATES **/
.item-container.page.dark:hover:not(.selected):not(:focus) {
  background-color: var(--ic-action-dark-bg-hover);
}

.item-container.page.dark:active:not(.selected):not(:focus) {
  background-color: var(--ic-action-dark-bg-active);
}

ic-typography.dark {
  color: var(--ic-color-primary-text);
}

.selected.dark::after {
  content: "";
  height: var(--ic-space-xxs);
  width: 100%;
  position: absolute;
  bottom: 0;
  background-color: var(--ic-action-dark);
}

.disabled {
  color: var(--ic-architectural-200);
  pointer-events: none;
  cursor: none;
  user-select: none;
}

@media (forced-colors: active) {
  .selected::after {
    background-color: Highlight;
  }

  .selected.light::after {
    background-color: Highlight;
  }

  .selected.dark::after {
    background-color: Highlight;
  }
}
