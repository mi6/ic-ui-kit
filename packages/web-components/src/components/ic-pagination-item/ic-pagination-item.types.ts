export type IcPaginationItemType = "page" | "ellipsis" | "simple-current";
