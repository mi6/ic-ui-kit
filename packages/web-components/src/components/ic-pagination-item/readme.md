# ic-pagination-item



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                                                                                               | Type                                       | Default     |
| -------------- | --------------- | --------------------------------------------------------------------------------------------------------- | ------------------------------------------ | ----------- |
| `appearance`   | `appearance`    | The appearance of the pagination, e.g. dark, light or the default.                                        | `"dark" \| "default" \| "light"`           | `undefined` |
| `ariaOverride` | `aria-override` | If true the aria-label will be set to `Page X of Y`, where X is the current page and Y is the page count. | `boolean`                                  | `false`     |
| `disabled`     | `disabled`      | If `true`, the pagination item will be disabled.                                                          | `boolean`                                  | `false`     |
| `label`        | `label`         | The label for the pagination item (applicable when simple pagination is being used).                      | `string`                                   | `"Page "`   |
| `page`         | `page`          | The current page number.                                                                                  | `number`                                   | `undefined` |
| `pages`        | `pages`         | The total number of pages.                                                                                | `number`                                   | `undefined` |
| `selected`     | `selected`      | If `true`, the pagination item will be selected.                                                          | `boolean`                                  | `false`     |
| `type`         | `type`          | The type of pagination item - 'page' or 'ellipsis'.                                                       | `"ellipsis" \| "page" \| "simple-current"` | `undefined` |


## Dependencies

### Used by

 - [ic-pagination](../ic-pagination)

### Depends on

- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-pagination-item --> ic-typography
  ic-pagination --> ic-pagination-item
  style ic-pagination-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


