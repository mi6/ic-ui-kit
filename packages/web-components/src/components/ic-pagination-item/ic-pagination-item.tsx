import {
  Component,
  Prop,
  State,
  Element,
  h,
  Event,
  EventEmitter,
  Watch,
} from "@stencil/core";
import { IcPaginationItemType } from "./ic-pagination-item.types";
import { IcThemeForeground } from "../../utils/types";
import { removeDisabledFalse, capitalize } from "../../utils/helpers";

@Component({
  tag: "ic-pagination-item",
  styleUrl: "ic-pagination-item.css",
  shadow: true,
})
export class PaginationItem {
  @Element() el: HTMLIcPaginationItemElement;

  @State() capitalizedLabel: string;

  /**
   * The appearance of the pagination, e.g. dark, light or the default.
   */
  @Prop() appearance: IcThemeForeground = "default";

  /**
   * If `true`, the pagination item will be disabled.
   */
  @Prop() disabled: boolean = false;

  /**
   * The label for the pagination item (applicable when simple pagination is being used).
   */
  @Prop() label: string = "Page";

  @Watch("label")
  watchLabelHandler(): void {
    this.capitalizedLabel = capitalize(this.label);
  }

  /**
   * The current page number.
   */
  @Prop() page: number | null;

  /**
   * If `true`, the pagination item will be selected.
   */
  @Prop() selected: boolean = false;

  /**
   * The type of pagination item - 'page' or 'ellipsis'.
   */
  @Prop() type: IcPaginationItemType;

  /**
   * @internal - Emitted when a pagination item is clicked.
   */
  @Event() paginationItemClick: EventEmitter<{ page: number }>;

  componentWillLoad(): void {
    this.watchLabelHandler();
    removeDisabledFalse(this.disabled, this.el);
  }

  private handleClick = () => {
    this.paginationItemClick.emit({ page: this.page });
  };

  render() {
    const {
      page,
      selected,
      type,
      disabled,
      appearance,
      label,
      capitalizedLabel,
    } = this;

    return (
      <a>
        {type === "ellipsis" ? (
          <div
            class={{
              ["item-container"]: true,
              ["ellipsis"]: true,
              ["disabled"]: disabled,
            }}
          >
            <ic-typography
              class={{
                [`${appearance}`]: true,
              }}
              variant="label"
            >
              ...
            </ic-typography>
          </div>
        ) : type === "page" ? (
          <button
            disabled={disabled}
            onClick={this.handleClick}
            tabindex={selected ? "-1" : "0"}
            role="button"
            aria-current={selected && "page"}
            aria-label={
              selected ? `${label}: ${page}` : `Go to ${label} ${page}`
            }
            class={{
              ["selected"]: !disabled && selected,
              ["disabled"]: disabled,
              ["item-container"]: true,
              [`${appearance}`]: true,
              ["page"]: type === "page",
            }}
          >
            <ic-typography
              variant="label"
              class={{
                ["page-selected"]: selected,
                ["disabled"]: disabled,
                [`${appearance}`]: true,
              }}
            >
              {page}
            </ic-typography>
          </button>
        ) : (
          <ic-typography
            aria-live="polite"
            class={{
              [`simple-current ${appearance}`]: true,
              ["disabled"]: disabled,
            }}
            variant="label"
          >
            {capitalizedLabel} {page}
          </ic-typography>
        )}
      </a>
    );
  }
}
