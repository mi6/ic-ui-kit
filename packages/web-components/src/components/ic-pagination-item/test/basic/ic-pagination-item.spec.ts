import { PaginationItem } from "../../ic-pagination-item";
import { newSpecPage } from "@stencil/core/testing";

describe("ic-pagination-item", () => {
  it("should render default pagination item", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render default pagination item");
  });

  it("should render with custom label", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1 label="screen"></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render with custom label");
  });

  it("should render with dark appearance", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1 type="page" appearance="dark"></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render with dark appearance");
  });

  it("should render with light appearance", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1 type="page" appearance="light"></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render with light appearance");
  });

  it("should render as selected item", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1 type="page" selected></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render as selected item");
  });

  it("should render disabled", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1 type="page" disabled></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render disabled");
  });

  it("should render as ellipsis type", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1 type="ellipsis"></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render as ellipsis type");
  });

  it("should render disabled ellipsis type", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1 type="ellipsis" disabled></ic-pagination-item>`,
    });

    expect(page.root).toMatchSnapshot("render disabled ellipsis type");
  });

  it("should test paginationItemClick event", async () => {
    const page = await newSpecPage({
      components: [PaginationItem],
      html: `<ic-pagination-item page=1></ic-pagination-item>`,
    });

    const eventSpy = jest.fn();

    page.root.addEventListener("paginationItemClick", eventSpy);

    page.rootInstance.handleClick();

    await page.waitForChanges();

    expect(eventSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: expect.objectContaining({
          page: 1,
        }),
      })
    );
  });
});
