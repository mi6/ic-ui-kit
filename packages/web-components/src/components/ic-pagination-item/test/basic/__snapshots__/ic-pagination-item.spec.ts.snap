// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-pagination-item should render as ellipsis type: render as ellipsis type 1`] = `
<ic-pagination-item page="1" type="ellipsis">
  <mock:shadow-root>
    <a>
      <div class="ellipsis item-container">
        <ic-typography class="default" variant="label">
          ...
        </ic-typography>
      </div>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;

exports[`ic-pagination-item should render as selected item: render as selected item 1`] = `
<ic-pagination-item page="1" selected="" type="page">
  <mock:shadow-root>
    <a>
      <button aria-current="page" aria-label="Page: 1" class="default item-container page selected" role="button" tabindex="-1">
        <ic-typography class="default page-selected" variant="label">
          1
        </ic-typography>
      </button>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;

exports[`ic-pagination-item should render default pagination item: render default pagination item 1`] = `
<ic-pagination-item page="1">
  <mock:shadow-root>
    <a>
      <ic-typography aria-live="polite" class="default simple-current" variant="label">
        Page 1
      </ic-typography>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;

exports[`ic-pagination-item should render disabled ellipsis type: render disabled ellipsis type 1`] = `
<ic-pagination-item disabled="" page="1" type="ellipsis">
  <mock:shadow-root>
    <a>
      <div class="disabled ellipsis item-container">
        <ic-typography class="default" variant="label">
          ...
        </ic-typography>
      </div>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;

exports[`ic-pagination-item should render disabled: render disabled 1`] = `
<ic-pagination-item disabled="" page="1" type="page">
  <mock:shadow-root>
    <a>
      <button aria-label="Go to Page 1" class="default disabled item-container page" disabled="" role="button" tabindex="0">
        <ic-typography class="default disabled" variant="label">
          1
        </ic-typography>
      </button>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;

exports[`ic-pagination-item should render with custom label: render with custom label 1`] = `
<ic-pagination-item label="screen" page="1">
  <mock:shadow-root>
    <a>
      <ic-typography aria-live="polite" class="default simple-current" variant="label">
        Screen 1
      </ic-typography>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;

exports[`ic-pagination-item should render with dark appearance: render with dark appearance 1`] = `
<ic-pagination-item appearance="dark" page="1" type="page">
  <mock:shadow-root>
    <a>
      <button aria-label="Go to Page 1" class="dark item-container page" role="button" tabindex="0">
        <ic-typography class="dark" variant="label">
          1
        </ic-typography>
      </button>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;

exports[`ic-pagination-item should render with light appearance: render with light appearance 1`] = `
<ic-pagination-item appearance="light" page="1" type="page">
  <mock:shadow-root>
    <a>
      <button aria-label="Go to Page 1" class="item-container light page" role="button" tabindex="0">
        <ic-typography class="light" variant="label">
          1
        </ic-typography>
      </button>
    </a>
  </mock:shadow-root>
</ic-pagination-item>
`;
